packages/arb-validator/rollupvalidator/txtracker.go

```diff
@@ -110,214 +123,238 @@ func (a *assertionInfo) FindLogs(address *big.Int, topics []common.Hash) []logRe
+	nodeData, err := tr.txDB.lookupNodeRecord(height, nodeHash)
+	if err != nil {
+		return nil, err
+	}
 
-	info.OutMessages = assertion.Assertion.OutMsgs
-	info.AssertNodeHash = assertion.NodeHash
-	info.LogsValHashes = make([]string, 0, len(logs))
-	info.LogsAccHashes = make([]string, 0, len(logs))
+	if msgIndex < 0 || msgIndex >= int64(len(nodeData.AVMMessages)) {
+		return nil, err
+	}
+	return nodeData.AVMMessages[msgIndex], nil
+}
 
-	acc := common.Hash{}
-	for _, logsVal := range logs {
-		logsValHash := logsVal.Hash()
-		info.LogsValHashes = append(info.LogsValHashes,
-			hexutil.Encode(logsValHash[:]))
-		acc = hashing.SoliditySHA3(
-			hashing.Bytes32(acc),
-			hashing.Bytes32(logsValHash),
-		)
-		info.LogsAccHashes = append(info.LogsAccHashes,
-			hexutil.Encode(acc.Bytes()))
+func (tr *txTracker) TxInfo(ctx context.Context, txHash common.Hash) (txInfo, error) {
+	tr.RLock()
+	defer tr.RUnlock()
+	select {
+	case <-ctx.Done():
+		return txInfo{Found: false}, errors.New("call timed out")
+	default:
+	}
+	tx, err := tr.txDB.lookupTxRecord(txHash)
+	if err != nil || tx == nil {
+		return txInfo{Found: false}, err
+	}
+	nodeInfo, err := tr.txDB.lookupNodeRecord(tx.NodeHeight, tx.NodeHash.Unmarshal())
+	if err != nil {
+		return txInfo{Found: false}, nil
 	}
+	return getTxInfo(txHash, nodeInfo, tx), nil
+}
 
-	var logsPostHash string
-	if len(logs) > 0 {
-		logsPostHash = info.LogsAccHashes[len(info.LogsAccHashes)-1]
-	} else {
-		logsPostHash = hexutil.Encode(zero[:])
+func (tr *txTracker) AssertionCount(ctx context.Context) (uint64, error) {
+	tr.RLock()
+	defer tr.RUnlock()
+	select {
+	case <-ctx.Done():
+		return 0, errors.New("call timed out")
+	default:
 	}
+	return tr.maxNodeHeight, nil
+}
 
-	for i, logVal := range logs {
-		if i > 0 {
-			logsPreHash = info.LogsAccHashes[i-1] // Previous acc hash
-		}
-		logsValHashes := info.LogsValHashes[i+1:] // log acc hashes after logVal
-
-		txInfo := txInfo{
-			Found:          true,
-			assertionIndex: len(tr.assertionInfo),
-			RawVal:         logVal,
-			LogsPreHash:    logsPreHash,
-			LogsPostHash:   logsPostHash,
-			LogsValHashes:  logsValHashes,
-			OnChainTxHash:  disputableTxHash,
+func (tr *txTracker) FindLogs(
+	ctx context.Context,
+	fromHeight *int64,
+	toHeight *int64,
+	address *common.Address,
+	topics []common.Hash,
+) ([]*validatorserver.LogInfo, error) {
+	tr.RLock()
+	defer tr.RUnlock()
+	select {
+	case <-ctx.Done():
+		return nil, errors.New("call timed out")
+	default:
+	}
+	startHeight := int64(0)
+	endHeight := int64(tr.maxNodeHeight)
+	if fromHeight != nil && *fromHeight > int64(0) {
+		startHeight = *fromHeight
+	}
+	if toHeight != nil {
+		altEndHeight := *toHeight + 1
+		if endHeight > altEndHeight {
+			endHeight = altEndHeight
 		}
+	}
+	logs := make([]*validatorserver.LogInfo, 0)
+	if startHeight >= int64(tr.maxNodeHeight) {
+		return logs, nil
+	}
 
-		evmVal, err := evm.ProcessLog(logVal, tr.vmID)
+	for i := startHeight; i < endHeight; i++ {
```

Comment:
If this loop might run for a long time, it would be good to check periodically whether ctx.Done has fired,  then return a timeout error if so.

Reply:
That's a good idea. This could definitely take a while to run